lib/ecto/adapters/pool/transaction.ex

```diff
@@ -0,0 +1,312 @@
+defmodule Ecto.Adapters.Pool.Transaction do
+  @moduledoc """
+  Behaviour and implementation for carrying out nested transactions using a
+  connection from a pool.
+  """
+
+  use Behaviour
+  alias Ecto.Adapters.Pool
+
+  @typedoc """
+  Opaque transaction reference.
+
+  Use inside `transaction/4` to retrieve the connection term or to disconnect
+  the transaction, closing the connection.
+  """
+  @opaque ref :: {__MODULE__, module, Pool.t}
+
+  @typedoc """
+  The mode of a transaction.
+
+  Supported modes:
+
+    * `:raw` - direct transaction without a sandbox
+    * `:sandbox` - transaction inside a sandbox
+  """
+  @type mode :: :raw | :sandbox
+
+  @typedoc """
+  The depth of nested transactions.
+  """
+  @type depth :: pos_integer
+
+  @typedoc """
+  The time in microseconds spent waiting for a connection from the pool.
+  """
+  @type queue_time :: non_neg_integer
+
+  @doc """
+  Checkout a connection from a pool and open a transaction.
+
+  Returns `{mode, conn, queue_time}` on success, where `conn` is any connection
+  data associated with the transaction. The connection data can be retrieved
+  inside a `transaction/4` with `connection/2`.
+
+  Returns `{:error, :noproc}` if the pool is not alive and
+  `{:error, :noconnect}` if a connection is not available.
+  """
+  defcallback open_transaction(Pool.t, timeout) ::
+    {mode, conn, queue_time} | {:error, :noproc | :noconnect} when conn: any
+
+  @doc """
+  Get the connection module and process from the connection data for the current
+  transaction.
+  """
+  defcallback transaction_connection(Pool.t, conn, timeout) ::
```

Comment:
I am assuming you need this function for sbroker? Otherwise, why wouldn't you simply return the connection as a result of open_transaction?